### PR TITLE
DDP-5592: Embedded Activity Block

### PR DESCRIPTION
**Pull request by EdPokhvalin**
This change is [Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/475)

### CODE REVIEW

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.html [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+<ng-template #section>
+  <ddp-activity-section [section]="activity.sections[0]"
+                        [studyGuid]="studyGuid"
+                        [activityGuid]="instance.instanceGuid"
+                        [readonly]="readonly"
+                        [validationRequested]="validationRequested">
+  </ddp-activity-section>
+
+  <button mat-button
+          *ngIf="singleSection">
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.html]: What does this button do?

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { EmbeddedActivityBlockComponent } from 'ddp-sdk';
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts]: It is in same project, so this import probably will result in circular reference error (have not checked yet). Try relative path instead.

Comment (MocanaAtBroad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts]: But if this is just the generated test and we are not writing a real one, then just delete. A test would be nice, though!

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+              private logger: LoggingService) { }
+
+  ngOnInit(): void {
+    this.getActivity();
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (changes.readonly.currentValue && changes.validationRequested.currentValue) { // if we just submitted the form
+      this.componentBusy.emit(true);
+      this.activityServiceAgent.flushForm(this.studyGuid, this.instance.instanceGuid)
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts]: This seems like a very indirect way to submit/flushForm. Still need to dig deeper on my end, but perhaps the button click should trigger this?

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+  private getActivity(): void {
+    this.activityServiceAgent.getActivity(of(this.studyGuid), of(this.instance.instanceGuid))
+      .pipe(
+        catchError(err => {
+          this.logger.logError(this.LOG_SOURCE, 'An error during getting a full activity', err);
+          return EMPTY;
+        }),
+        take(1))
+      .subscribe(activity => {
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts]: Code looks correct to me, but my IDE TypeScript compiler is complaining about activity type being "unknown", so to make it happy:
```suggestion
     .subscribe((activity: ActivityForm) => {
```

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityBlock/activityBlock.component.html [hunk 5]
```diff
@@ -13,6 +13,16 @@
                 </ddp-modal-activity-block>
             </div>
         </ng-container>
+        <ng-template #embedded>
+          <div *ngFor="let instance of childInstances; trackBy: getInstanceId" class="card">
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityBlock/activityBlock.component.html]: You have this line twice. I suggest simplifying it and using 
`*ngIf="isModal; else embedded"` conditional inside of the *ngFor loop

**Comment by alextanlan** [comment]
Tests fail in your branch
image

**Comment by alextanlan** [comment]
Did you consider how it should look like in case of having more than 1 embedded component on a page (e.g. Parents page:  mother + father).
For now it looks like a list with combined questions for all instances (without any separators or labels/titles).
See below


image

**Review by EdPokhvalin — COMMENTED** [review]
The circular dependency has been fixed. As for your second comment, what do you mean by "separators or labels/titles"? In single-page mode (for now, at least) each embedded block will have it's own "*Submit*" button, it will separate the blocks from each other.

> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-:-MW9ICZR6Z0u1uk6r9Ij:b-dkcuz5)* status: 0 of 9 files reviewed, 5 unresolved discussions (waiting on @alextanlan, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityBlock/activityBlock.component.html, line 17 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MVvF4ij5nal2o97Wi5t-r1-17:-MW9I9R99nKfakzDx8oo:bb6un1j) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/ef6dcbae4ab083d3fd5e1d1685a296f35fd28e4b/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityBlock/activityBlock.component.html#L17)):*
<details><summary><i>Previously, alextanlan (Alex Maratkanov) wrote…</i></summary><blockquote>

You have this line twice. I suggest simplifying it and using 
`*ngIf="isModal; else embedded"` conditional inside of the *ngFor loop
</blockquote></details>

Done
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.html, line 14 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MVvF4ij5nal2o97Wi5v-r1-14:-MW9LbvbDsKfF3_c0kvZ:boc109f) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/ef6dcbae4ab083d3fd5e1d1685a296f35fd28e4b/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.html#L14)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

What does this button do?
</blockquote></details>

Forgot about these changes too. This button used in single-page mode where we need to submit a child activity separately from the parent one
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts, line 45 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MVvF4ij5nal2o97Wi5y-r1-45:-MW9J0Qs3xna-I7DSvyO:bux9q5l) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/ef6dcbae4ab083d3fd5e1d1685a296f35fd28e4b/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts#L45)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

This seems like a very indirect way to submit/flushForm. Still need to dig deeper on my end, but perhaps the button click should trigger this?
</blockquote></details>

I reworked this piece of functionality.
Now we can use this component in two modes: as a part of the wizard and on a single page:

1. In the wizard we must complete this activity only when "*Submit*" button was pressed. In that case we just track changes in *readonly* and *validationRequested* properties. This condition is my approach to how we can understand that "Submit" button was clicked. I was unable to figure out a different way. 
2. In the single page mode we should have our own "*Submit*" button. The activity will be completed only when we'll press this button.

I also don't see any possibilities to resolve the mode type inside of the component, so I added "*singleSection*" property
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts, line 64 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MVvF4ij5nal2o97Wi5y-r1-64:-MW9I7ZP4nBY62HX7tc8:b-886gsx) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/ef6dcbae4ab083d3fd5e1d1685a296f35fd28e4b/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts#L64)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Code looks correct to me, but my IDE TypeScript compiler is complaining about activity type being "unknown", so to make it happy:
```suggestion
     .subscribe((activity: ActivityForm) => {
```
</blockquote></details>

Fixed
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts, line 3 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MVvF4ij5nal2o97Wi5x-r1-3:-MW9M-q04rCf7L-AZ9wK:b-b4rxsg) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/ef6dcbae4ab083d3fd5e1d1685a296f35fd28e4b/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts#L3)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

But if this is just the generated test and we are not writing a real one, then just delete. A test would be nice, though!
</blockquote></details>

I have fixed the circular dependency and will try to make an appropriate test

**Review by MocanaAtBroad — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/475)* status: 0 of 11 files reviewed, 11 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity.component.ts, line 234 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWR3O-tC4xOFJ4osx_a:-MWR3O-tC4xOFJ4osx_b:b-3raz2n) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/9142c4f803866f2a7cfa69518eaf81bdd7857c46/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity.component.ts#L234)):*
> ```XML
>     public ngOnInit(): void {
>         this.getActivity();
>         const submitSub = this.submitAttempted.subscribe(response => this.submitService.announceSubmit(null));
> ```

We are going to use the initial indication that a submission will be attempted to trigger the announceSubmit.
So when user presses button to submit the activity, first thing is it will tell all listeners of the submitService that we are submitting.
(before we were waiting until AFTER the parent activity submit was completed!).
ActivityBlock will listen for this announcement to submit the child activity instance.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/baseActivity.component.ts, line 40 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWR4-EA97FKgZkKop9d:-MWR4-EA97FKgZkKop9e:bdwb06d) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/9142c4f803866f2a7cfa69518eaf81bdd7857c46/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/baseActivity.component.ts#L40)):*
> ```XML
>     @Input() studyGuid: string;
>     @Input() activityGuid: string;
>     @Output() submit: EventEmitter<ActivityResponse | null> = new EventEmitter(true);
> ```

Want these events to be async
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityBlock/activityBlock.component.ts, line 23 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWR47Lk40OJo-4ds5t4:-MWR47LlAj-3grzhy1Wf:b-bsn7ux) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/9142c4f803866f2a7cfa69518eaf81bdd7857c46/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityBlock/activityBlock.component.ts#L23)):*
> ```XML
>     @Input() studyGuid: string;
>     @Input() parentActivityInstanceGuid: string;
>     @Output() embeddedComponentBusy: EventEmitter<boolean> = new EventEmitter<boolean>(true);
> ```

async again.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.html, line 14 at r1](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MVvF4ij5nal2o97Wi5v-r1-14:-MWR4BAO5Kr5xvIWRvsi:b-saarof) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/ef6dcbae4ab083d3fd5e1d1685a296f35fd28e4b/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.html#L14)):*
<details><summary><i>Previously, EdPokhvalin wrote…</i></summary><blockquote>

Forgot about these changes too. This button used in single-page mode where we need to submit a child activity separately from the parent one
</blockquote></details>

In my update I just removed button. Was not doing anything. We can bring it back if we ever need it.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts, line 29 at r2](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWPcEbD5mVMV-zb3iFJ:-MWPcEbE9fXs41OYzUnM:b149gxa) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/370649abe730b4ea357e3ee1b86812daefe4ce86/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts#L29)):*
> ```XML
>   @Input() validationRequested: boolean;
>   @Input() studyGuid: string;
>   @Input() singleSection: boolean;
> ```

Not seeing `singleSection` being set as input parameter.
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts, line 46 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWR4Ssu70xK9eMWEytm:-MWR4SsvCws4Pjfggd17:b-7crvk3) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/9142c4f803866f2a7cfa69518eaf81bdd7857c46/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts#L46)):*
> ```XML
>     ).subscribe();
>   }
> ```

I refactored. the "flushForm" is triggered by listening for submitAnnounce event. We wrap the call of flushForm with busy/true false so parent activity knows to wait for final parent flush form (I think in Ed's original he was detecting the same thing, but very indirectly).
___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts, line 50 at r3](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWR4xXUCUKy-wh6XHkj:-MWR4xXVEPkUUSn5PhD_:bmw26ij) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/9142c4f803866f2a7cfa69518eaf81bdd7857c46/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.spec.ts#L50)):*
> ```XML
>     };
>     fixture.detectChanges();
>   });
> ```

Cleaned up test so it executes. Just checks that component can be instantiated.

**Review by MocanaAtBroad — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/475)* status: 0 of 11 files reviewed, 10 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts, line 29 at r2](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWPcEbD5mVMV-zb3iFJ:-MWdhypP5NPGJ9y3k6r5:b-8ae8o0) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/370649abe730b4ea357e3ee1b86812daefe4ce86/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts#L29)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Not seeing `singleSection` being set as input parameter.
</blockquote></details>

Please remove singleSection. Not used.

**Review by EdPokhvalin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/broadinstitute/ddp-angular/475)* status: 0 of 11 files reviewed, 9 unresolved discussions (waiting on @alextanlan, @EdPokhvalin, @fat23cat, and @MocanaAtBroad)

___
*[ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts, line 29 at r2](https://reviewable.io/reviews/broadinstitute/ddp-angular/475#-MWPcEbD5mVMV-zb3iFJ:-MWUJ3VF5ooKMuDY6_j6:bb6un1j) ([raw file](https://github.com/broadinstitute/ddp-angular/blob/370649abe730b4ea357e3ee1b86812daefe4ce86/ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/embeddedActivityBlock/embeddedActivityBlock.component.ts#L29)):*
<details><summary><i>Previously, MocanaAtBroad wrote…</i></summary><blockquote>

Please remove singleSection. Not used.
</blockquote></details>

Done

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**